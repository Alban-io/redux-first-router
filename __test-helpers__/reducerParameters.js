import createSmartHistory from '../src/smart-history'

import { createInitialState } from '../src/createLocationReducer'

export default async (type, pathname) => {
  // eslint-disable-line import/prefer-default-export
  const history = createSmartHistory({ initialEntries: ['/first'] })
  history.listen(function() {})
  await history.push(pathname)

  const current = { pathname, type, payload: { param: 'bar' } }
  const prev = { pathname: '/first', type: 'FIRST', payload: {} }
  const routesMap = {
    FIRST: { path: '/first' },
    SECOND: { path: '/second/:param' }
  }

  return {
    type,
    pathname,
    current,
    prev,

    initialState: createInitialState(
      routesMap,
      history
    ),

    routesMap,
    history,

    action: {
      type,
      payload: { param: 'bar' },
      meta: {
        location: {
          current,
          prev,
          kind: 'load',
          history: {
            entries: history.entries.slice(0), // history.entries.map(entry => entry.pathname)
            index: history.index,
            length: history.length
          }
        }
      }
    },

    expectState(state) {
      expect(state.pathname).toEqual(pathname)
      expect(state.type).toEqual(type)
      expect(state.payload).toEqual({ param: 'bar' })
      expect(state.prev).toEqual(prev)
      expect(state.kind).toEqual('load')

      expect(state).toMatchSnapshot()
    }
  }
}
