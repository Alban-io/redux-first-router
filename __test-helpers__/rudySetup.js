import { applyMiddleware, createStore, compose, combineReducers } from 'redux'
import reduxThunk from 'redux-thunk'
import createRouter from '../src/createRouter'
import createScene from '../src/createScene'
import { NOT_FOUND } from '../src/index'

import fakeAsyncWork from '../__test-helpers__/fakeAsyncWork'

export default (path = '/first', options = {}, custom = {}) => {
  const routesMap = {
    ...defaultRoutes,
    ...custom.routesMap
  }

  options.initialEntries = [path]
  options.extra = { arg: 'extra-arg' }

  const { types, actions, routes, exportString } = createScene(routesMap, custom)
  const { middleware, reducer, firstRoute, history } = createRouter(
    routes,
    options
  )

  const title = (state = {}, action = {}) => action.type
  const rootReducer = combineReducers({ title, location: reducer })
  const enhancer = applyMiddleware(middleware, reduxThunk)
  const store = createStore(rootReducer, enhancer)

  return {
    store,
    firstRoute,
    history,
    types,
    actions,
    exportString,
    routes
  }
}

export const defaultRoutes = {
  FIRST: '/first',
  SECOND: '/second',
  THIRD: {
    path: '/third',
    thunk: async () => {
      await fakeAsyncWork()
      return 'thunk'
    }
  },
  FOURTH: {
    path: '/fourth',
    thunk: async () => {
      await fakeAsyncWork()
      return 'thunk'
    },
    onComplete: () => {
      return 'onComplete'
    },
    action: (arg) => (req, type) => {
      return { payload: { foo: arg }, type }
    }
  },
  PLAIN: {
    action: (arg) => {
      return { foo: arg }
    }
  },
  [NOT_FOUND]: '/not-found-foo'
}

export const log = store => {
  const state = store.getState().location
  delete state.routesMap
  delete state.hasSSR
  console.log(state)
}
