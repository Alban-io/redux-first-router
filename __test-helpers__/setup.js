import createHistory from 'history/createMemoryHistory'
import connectRoutes from '../src/connectRoutes'


export default (path = '/', options = { title: 'title', location: 'location' }) => {
  const routesMap = {
    FIRST: '/first',
    SECOND: '/second/:param',
  }

  const history = createHistory({
    initialEntries: [path],
    initialIndex: 0,
    keyLength: 6,
  })

  return connectRoutes(history, routesMap, options)
}
