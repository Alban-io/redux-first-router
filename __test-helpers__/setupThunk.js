import { createStore, applyMiddleware, compose } from 'redux'
import createHistory from 'history/createMemoryHistory'
import connectTypes from '../src/connectTypes'


export default (path = '/', thunk) => {
  const routesMap = {
    FIRST: '/first',
    SECOND: { path: '/second/:param', thunk },
    THIRD: { path: '/third/:param' },
  }

  const history = createHistory({
    initialEntries: [path],
    initialIndex: 0,
    keyLength: 6,
  })

  const { middleware, enhancer, thunk: ssrThunk, reducer: locationReducer } = connectTypes(history, routesMap)

  const middlewares = applyMiddleware(middleware)

  const reducer = (state = {}, action = {}) => ({
    location: locationReducer(state.location, action),
  })

  const store = createStore(reducer, undefined, compose(enhancer, middlewares))

  return { store, thunk: ssrThunk }
}
