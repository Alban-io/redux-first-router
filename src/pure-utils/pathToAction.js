// @flow
import pathToRegexp from 'path-to-regexp'
import { NOT_FOUND } from '../actions'
import type { Routes, RouteNames, PlainAction as Action } from '../flow-types'


export default (
  path: string,
  routes: Routes,
  routeNames: RouteNames,
): Action => {
  let i = 0
  let match
  const keys = []

  while (!match && i < routes.length) {
    keys.length = 0 // empty the array and start over
    const routePath = routes[i].path || routes[i] // route may be an object containing a route or a route string itself
    const reg = pathToRegexp(routePath, keys)
    match = reg.exec(path)
    i++
  }

  if (match) {
    i--

    const capitalizedWords = typeof routes[i] === 'object' && routes[i].capitalizedWords
    const fromPath = routes[i] && typeof routes[i].fromPath === 'function' && routes[i].fromPath
    const type = routeNames[i]

    const payload = keys.reduce((payload, key, index) => {
      let value = match && match[index + 1] // item at index 0 is the overall match, whereas those after correspond to the key's index

      value = !isNaN(value)
        ? parseFloat(value) // make sure pure numbers aren't passed to reducers as strings
        : value

      value = capitalizedWords && typeof value === 'string'
        ? value.replace(/-/g, ' ').replace(/\b\w/g, l => l.toUpperCase()) // 'my-category' -> 'My Category'
        : value

      value = fromPath && typeof value === 'string'
        ? fromPath(value, key.name)
        : value

      payload[key.name] = value

      return payload
    }, {})

    return { type, payload }
  }

  // This will basically will only end up being called if the developer is manually calling history.push().
  // Or, if visitors visit an invalid URL, the developer can use the NOT_FOUND type to show a not-found page to
  return { type: NOT_FOUND, payload: {} }
}
