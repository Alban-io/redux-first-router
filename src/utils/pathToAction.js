// @flow
import qs from 'query-string'
import { compilePath } from './matchPath'
import { stripBasename, parsePath } from '../smart-history/utils/path'
import notFound from '../action-creators/notFound'

import type { RoutesMap, ReceivedAction } from '../flow-types'

export default (
  loc: Object | string,
  routes: RoutesMap,
  basename: string = ''
): ReceivedAction => {
  const { url, state } = typeof loc === 'string' ? { url: loc } : loc
  const { pathname, search, hash } = parsePath(url)
  const path = basename ? stripBasename(pathname, basename) : pathname
  const query = search ? qs.parse(search) : {}
  const types = Object.keys(routes).filter(type => routes[type].path)

  let i = 0
  let match
  let keys
  let route
  let type

  while (!match && i < types.length) {
    type = types[i]
    route = routes[type]

    const result = isPathMatch(path, route.path)
    match = result.match
    keys = result.keys

    if (match && route.query && !isQueryMatch(query, route.query)) {
      match = null
    }

    if (match && route.hash && !isMatch(hash, route.hash)) {
      match = null
    }

    i++
  }

  if (match && route) {
    const { capitalizedWords, fromPath } = route

    const payload = (keys || []).reduce((payload, key, index) => {
      let val = match && match[index + 1] // item at index 0 is the overall match, whereas those after correspond to the key's index

      if (typeof val === 'string') {
        if (fromPath) {
          val = fromPath && fromPath(val, key.name)
        }
        else if (isNumber(val)) {
          val = parseFloat(val)
        }
        else if (capitalizedWords) {
          val = val.replace(/-/g, ' ').replace(/\b\w/g, l => l.toUpperCase()) // 'my-category' -> 'My Category'
        }
      }

      payload[key.name] = val
      return payload
    }, {})


    return { type, payload, query, hash, state }
  }

  // This will basically only end up being called if the developer is manually calling history.push().
  // Or, if visitors visit an invalid URL, the developer can use the NOT_FOUND type to show a not-found page to
  return notFound({ query, hash, state }, url)
}


const isNumber = (val: string) => !val.match(/^\s*$/) && !isNaN(val)

const isPathMatch = (path, matcher) => {
  const { re, keys } = compilePath(matcher)
  const match = re.exec(path)
  return { match, keys }
}

const isQueryMatch = (query, matcher) => {
  for (const key in matcher) {
    const val = query[key]
    const expected = matcher[key]
    if (!isMatch(val, expected)) return false
  }

  return true
}

const isMatch = (val, expected) => {
  const type = typeof expected

  if (type === 'boolean') {
    return val !== '' && val !== undefined
  }
  else if (type === 'string') {
    return expected === val
  }
  else if (type === 'function') {
    return expected(val)
  }
  else if (expected instanceof RegExp) {
    return expected.test(val)
  }

  return true
}
