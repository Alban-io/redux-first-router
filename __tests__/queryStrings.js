import querySerializer from 'query-string'

import { setupAll } from '../__test-helpers__/setup'
import pathToAction from '../src/pure-utils/pathToAction'

it('dispatched as action.query', async () => {
  const { store } = await setupAll('/third', { querySerializer })
  const query = { foo: 'bar', baz: 69 }

  await store.dispatch({ type: 'FIRST', query })
  await store.dispatch({ type: 'THIRD', query })

  const state = store.getState() /*? $.location */
  expect(state).toMatchSnapshot()
})

it('dispatched as action.meta.query', async () => {
  const { store } = await setupAll('/third', { querySerializer })
  const query = { foo: 'bar', baz: 69 }

  await store.dispatch({ type: 'FIRST', meta: { query } })
  await store.dispatch({ type: 'THIRD', meta: { query } })

  const state = store.getState() /*? $.location */
  expect(state).toMatchSnapshot()
})

it('dispatched as action.payload.query', async () => {
  const { store } = await setupAll('/third', { querySerializer })
  const payload = { foo: 'bar', baz: 69 }

  await store.dispatch({ type: 'FIRST', meta: { payload } })
  await store.dispatch({ type: 'THIRD', meta: { payload } })

  const state = store.getState() /*? $.location */
  expect(state).toMatchSnapshot()
})

it('history.push("/path?search=foo")', async () => {
  const { store, history } = await setupAll('/third', { querySerializer })

  await history.push('/first?foo=bar&baz=69')
  await history.push('/third?foo=bar&baz=69')

  const state = store.getState() /*? $.location */
  expect(state).toMatchSnapshot()
})

it('currentPathName changes, but pathname stays the same (only query changes)', async () => {
  const { store, history } = await setupAll('/third', { querySerializer })

  await history.push('/first?foo=bar&baz=69')
  await history.push('/first?foo=car&baz=70')

  const state = store.getState() /*? $.location */
  expect(state).toMatchSnapshot()
})

it('generated from pathToAction within <Link />', async () => {
  const { store, history, routesMap } = await setupAll('/third', { querySerializer })

  let action = pathToAction(
    '/first?foo=bar&baz=69',
    routesMap,
    querySerializer
  ) /*? */
  await store.dispatch(action)

  action = pathToAction(
    '/first?foo=car&baz=70',
    routesMap,
    querySerializer
  ) /*? */
  await store.dispatch(action)

  const state = store.getState() /*? $.location */
  expect(state).toMatchSnapshot()
})
