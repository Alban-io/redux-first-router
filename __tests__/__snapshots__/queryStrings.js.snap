// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`currentPathName changes, but pathname stays the same (only query changes) 1`] = `
Object {
  "location": Object {
    "basename": "",
    "entries": Array [
      Object {
        "hash": "",
        "key": "345678",
        "pathname": "/third",
        "search": "",
        "state": Object {},
        "url": "/third",
      },
      Object {
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "?foo=bar&baz=69",
        "state": Object {},
        "url": "/first?foo=bar&baz=69",
      },
      Object {
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "?foo=car&baz=70",
        "state": Object {},
        "url": "/first?foo=car&baz=70",
      },
    ],
    "hasSSR": undefined,
    "index": 2,
    "kind": "push",
    "length": 3,
    "pathname": "/first",
    "payload": Object {},
    "prev": Object {
      "index": 1,
      "kind": "push",
      "length": 2,
      "pathname": "/first",
      "payload": Object {},
      "query": Object {
        "baz": "69",
        "foo": "bar",
      },
      "type": "FIRST",
    },
    "query": Object {
      "baz": "70",
      "foo": "car",
    },
    "search": "foo=car&baz=70",
    "type": "FIRST",
    "url": "/first?foo=car&baz=70",
  },
  "title": "FIRST",
}
`;

exports[`dispatched as action.payload.query 1`] = `
Object {
  "location": Object {
    "basename": "",
    "entries": Array [
      Object {
        "hash": "",
        "key": "345678",
        "pathname": "/third",
        "search": "",
        "state": Object {},
        "url": "/third",
      },
      Object {
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "",
        "state": Object {},
        "url": "/first",
      },
    ],
    "hasSSR": undefined,
    "index": 0,
    "kind": "back",
    "length": 2,
    "pathname": "/third",
    "payload": Object {},
    "prev": Object {
      "index": 1,
      "kind": "push",
      "length": 2,
      "pathname": "/first",
      "payload": Object {},
      "type": "FIRST",
    },
    "type": "THIRD",
    "url": "/third",
  },
  "title": "THIRD",
}
`;

exports[`generated from pathToAction within <Link /> 1`] = `
Object {
  "location": Object {
    "basename": "",
    "entries": Array [
      Object {
        "hash": "",
        "key": "345678",
        "pathname": "/third",
        "search": "",
        "state": Object {},
        "url": "/third",
      },
      Object {
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "?baz=69&foo=bar",
        "state": Object {},
        "url": "/first?baz=69&foo=bar",
      },
      Object {
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "?baz=70&foo=car",
        "state": Object {},
        "url": "/first?baz=70&foo=car",
      },
    ],
    "hasSSR": undefined,
    "index": 2,
    "kind": "push",
    "length": 3,
    "pathname": "/first",
    "payload": Object {},
    "prev": Object {
      "index": 1,
      "kind": "push",
      "length": 2,
      "pathname": "/first",
      "payload": Object {},
      "query": Object {
        "baz": "69",
        "foo": "bar",
      },
      "type": "FIRST",
    },
    "query": Object {
      "baz": "70",
      "foo": "car",
    },
    "search": "baz=70&foo=car",
    "type": "FIRST",
    "url": "/first?baz=70&foo=car",
  },
  "title": "FIRST",
}
`;

exports[`history.push("/path?search=foo") 1`] = `
Object {
  "location": Object {
    "basename": "",
    "entries": Array [
      Object {
        "hash": "",
        "key": "345678",
        "pathname": "/third",
        "search": "",
        "state": Object {},
        "url": "/third",
      },
      Object {
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "?foo=bar&baz=69",
        "state": Object {},
        "url": "/first?foo=bar&baz=69",
      },
      Object {
        "hash": "",
        "key": "345678",
        "pathname": "/third",
        "search": "?foo=bar&baz=69",
        "state": Object {},
        "url": "/third?foo=bar&baz=69",
      },
    ],
    "hasSSR": undefined,
    "index": 2,
    "kind": "push",
    "length": 3,
    "pathname": "/third",
    "payload": Object {},
    "prev": Object {
      "index": 1,
      "kind": "push",
      "length": 2,
      "pathname": "/first",
      "payload": Object {},
      "query": Object {
        "baz": "69",
        "foo": "bar",
      },
      "type": "FIRST",
    },
    "query": Object {
      "baz": "69",
      "foo": "bar",
    },
    "search": "foo=bar&baz=69",
    "type": "THIRD",
    "url": "/third?foo=bar&baz=69",
  },
  "title": "THIRD",
}
`;
